test/docker-e2e/e2e_state_sync_test.go

```diff
@@ -214,3 +210,423 @@ func (s *CelestiaTestSuite) TestStateSyncMocha() {
 
 	s.Require().NoError(err, "failed to wait for state sync to complete")
 }
+
+// TestStateSyncCompatibilityAcrossUpgrade verifies that a full node can state-sync to the latest height
+// when the chain history includes an on-chain upgrade.
+func (s *CelestiaTestSuite) TestStateSyncCompatibilityAcrossUpgrade() {
+	t := s.T()
+	if testing.Short() {
+		t.Skip("skipping state sync compatibility across upgrade test in short mode")
+	}
+
+	ctx := context.Background()
+
+	tag, err := dockerchain.GetCelestiaTagStrict()
+	s.Require().NoError(err)
+
+	t.Log("Phase 1: Starting v4 chain with 4 validators")
+	cfg := dockerchain.DefaultConfig(s.client, s.network).WithTag(tag)
+
+	var (
+		baseAppVersion   = appconsts.Version - 1
+		targetAppVersion = appconsts.Version
+	)
+	t.Logf("Starting chain with app version %d -> %d", baseAppVersion, targetAppVersion)
```

Comment:
[nit] this log is misleading because the chain can't start with app version "4 -> 5". Maybe

```suggestion
	t.Logf("Starting chain with app version %d and will upgrade to %d.", baseAppVersion, targetAppVersion)
```